File: v2.x/Examples/SciChartDemo/SciChartDemo/Views/TooltipsAndHitTest/ColumnDrillDownView.m
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2018. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// ColumnDrillDownView.m is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

#import "ColumnDrillDownView.h"
#import <SciChart/SciChart.h>

#pragma mark Custom modifier

@interface ColumnHitTest : SCIGestureModifier

@property (nonatomic, copy) SCIActionBlock_Pint tapedAtIndex;
@property (nonatomic, copy) SCIActionBlock doubleTaped;

@end

@implementation ColumnHitTest {
    CGPoint _location; // position of tap gesture
    int _index; // index of data column
}

- (BOOL)onTapGesture:(UITapGestureRecognizer*)gesture At:(UIView*)view {
    CGPoint location = [gesture locationInView:view];
    id<SCIRenderSurfaceProtocol> rs = self.parentSurface.renderSurface;
    if (![rs isPointWithinBounds:location]) return NO;
    if (gesture.state == UIGestureRecognizerStateEnded) {
        // save location of touch
        _location = location;
        id<SCIChartSurfaceProtocol> parent = self.parentSurface;
        SCIRenderableSeriesCollection * series = parent.renderableSeries;
        id<SCIRenderSurfaceProtocol> surface = parent.renderSurface;
        CGPoint actualLocation = [surface pointInChartFrame:_location];
        
        int count = (int)[series count];
        // check every renderable series for hit
        for (int i = 0; i < count; i++) {
            id<SCIRenderableSeriesProtocol> rSeries = [series itemAt:i];
            id<SCIRenderPassDataProtocol> data = [rSeries currentRenderPassData];
            id<SCIHitTestProviderProtocol> hitTest = [rSeries hitTestProvider]; // get hit test tools
            if (hitTest == nil) continue;
            
            // hit test verticaly: check if vertical projection through touch location crosses chart
            SCIHitTestInfo hitTestResult = [hitTest hitTestVerticalAtX:actualLocation.x Y:actualLocation.y Radius:5 onData:data];
            if (hitTestResult.match) {
                // if hit is registered on series
                // get values at closest point to hit test position
                _index = hitTestResult.index;
                if (_tapedAtIndex != nil) _tapedAtIndex(_index);
                break;
            }
        }
    }
    return YES;
}

- (BOOL)onDoubleTapGesture:(UITapGestureRecognizer *)gesture At:(UIView *)view {
    CGPoint location = [gesture locationInView:view];
    id<SCIRenderSurfaceProtocol> rs = self.parentSurface.renderSurface;
    if (![rs isPointWithinBounds:location]) return NO;
    if (_doubleTaped != nil) _doubleTaped();
    return YES;
}

@end

#pragma mark Palette provider

@interface ColumnDrillDownPalette : SCIPaletteProvider
@end

@implementation ColumnDrillDownPalette {
    NSMutableArray * _styles;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _styles = [NSMutableArray new];
    }
    return self;
}

- (void)addStyle:(id<SCIStyleProtocol>)style {
    [_styles addObject:style];
}

- (id<SCIStyleProtocol>)styleForX:(double)x Y:(double)y Index:(int)index {
    if (index > [_styles count]) return nil;
    return _styles[index];
}

@end

#pragma mark Example view

@implementation ColumnDrillDownView {
    SCIXyDataSeries * _totalData;
    
    SCIFastColumnRenderableSeries * _firstColumn;
    SCIFastColumnRenderableSeries * _secondColumn;
    SCIFastColumnRenderableSeries * _thirdColumn;
    SCIFastColumnRenderableSeries * _totalColumn;
    
    BOOL _isShowingTotal;
}

- (void)initExample {
    id<SCIAxis2DProtocol> xAxis = [SCICategoryNumericAxis new];
    xAxis.growBy = [[SCIDoubleRange alloc] initWithMin:SCIGeneric(0.1) Max:SCIGeneric(0.1)];
    
    id<SCIAxis2DProtocol> yAxis = [SCINumericAxis new];
    yAxis.growBy = [[SCIDoubleRange alloc] initWithMin:SCIGeneric(0.1) Max:SCIGeneric(0.1)];

    _totalData = [[SCIXyDataSeries alloc] initWithXType:SCIDataType_Int32 YType:SCIDataType_Int32];

    _firstColumn = [SCIFastColumnRenderableSeries new];
    _firstColumn.fillBrushStyle = [[SCILinearGradientBrushStyle alloc] initWithColorCodeStart:0xFF00FFFF finish:0xA000FFFF direction:SCILinearGradientDirection_Vertical];
    _firstColumn.strokeStyle = nil;
    _firstColumn.dataSeries = [self getDataSeries];
    
    _secondColumn = [SCIFastColumnRenderableSeries new];
    _secondColumn.fillBrushStyle = [[SCILinearGradientBrushStyle alloc] initWithColorCodeStart:0xFF00FF00 finish:0xA000FF00 direction:SCILinearGradientDirection_Vertical];
    _secondColumn.strokeStyle = nil;
    _secondColumn.dataSeries = [self getDataSeries];
    
    _thirdColumn = [SCIFastColumnRenderableSeries new];
    _thirdColumn.fillBrushStyle = [[SCILinearGradientBrushStyle alloc] initWithColorCodeStart:0xFFFF0000 finish:0xA0FF0000 direction:SCILinearGradientDirection_Vertical];
    _thirdColumn.strokeStyle = nil;
    _thirdColumn.dataSeries = [self getDataSeries];
    
     ColumnDrillDownPalette * palette = [ColumnDrillDownPalette new];
     [palette addStyle:[_firstColumn.style copy]];
     [palette addStyle:[_secondColumn.style copy]];
     [palette addStyle:[_thirdColumn.style copy]];
     
    _totalColumn = [SCIFastColumnRenderableSeries new];
    _totalColumn.fillBrushStyle = [[SCILinearGradientBrushStyle alloc] initWithColorCodeStart:0xFF505050 finish:0xA550005 direction:SCILinearGradientDirection_Vertical];
    _totalColumn.strokeStyle = nil;
    _totalColumn.dataSeries = _totalData;
    _totalColumn.paletteProvider = palette;
    
    ColumnHitTest * drillDownModifier = [ColumnHitTest new];
    __weak typeof(self) wSelf = self;
    drillDownModifier.doubleTaped = ^() { [wSelf showSeries:_totalColumn isTotal:YES]; };
    drillDownModifier.tapedAtIndex = ^(int index) { [wSelf showDetailedChartAt:index]; };
    
    [SCIUpdateSuspender usingWithSuspendable:self.surface withBlock:^{
        [self.surface.xAxes add:xAxis];
        [self.surface.yAxes add:yAxis];
        self.surface.chartModifiers = [[SCIChartModifierCollection alloc] initWithChildModifiers:@[[SCIPinchZoomModifier new], drillDownModifier]];
        
        SCIWaveRenderableSeriesAnimation * animation = [[SCIWaveRenderableSeriesAnimation alloc] initWithDuration:3 curveAnimation:SCIAnimationCurve_EaseOut];
        [animation setRepeatable:YES];
        
        [_firstColumn addAnimation:animation];
        [_secondColumn addAnimation:animation];
        [_thirdColumn addAnimation:animation];
        [_totalColumn addAnimation:animation];
        
        [self showSeries:_totalColumn isTotal:YES];
    }];
}

- (SCIXyDataSeries *)getDataSeries {
    SCIXyDataSeries * dataSeries = [[SCIXyDataSeries alloc] initWithXType:SCIDataType_Int32 YType:SCIDataType_Int32];
    
    int total = 0;
    int count = arc4random_uniform(5) + 3;
    for (int i = 0; i < count; i++) {
        int value = arc4random_uniform(5 + i) + 1;
        total += value;
        [dataSeries appendX:SCIGeneric(i) Y:SCIGeneric(value)];
    }
    [_totalData appendX:SCIGeneric(3) Y:SCIGeneric(total)];
    
    return dataSeries;
}

- (void)showDetailedChartAt:(int)index {
    if (!_isShowingTotal) return;
    if (index == 0) {
        [self showSeries:_firstColumn isTotal:NO];
    } else if (index == 1) {
        [self showSeries:_secondColumn isTotal:NO];
    } else if (index == 2) {
        [self showSeries:_thirdColumn isTotal:NO];
    }
}

- (void)showSeries:(id<SCIRenderableSeriesProtocol>)series isTotal:(BOOL)isTotal {
    _isShowingTotal = isTotal;
    [self.surface.renderableSeries clear];
    [self.surface.renderableSeries add:series];
    
    [self.surface.viewportManager zoomExtents];
}

@end

File: v2.x/Examples/SciChartSwiftDemo/SciChartSwiftDemo/Views/TooltipsAndHitTest/ColumnDrillDownView.swift
//******************************************************************************
// SCICHART® Copyright SciChart Ltd. 2011-2018. All rights reserved.
//
// Web: http://www.scichart.com
// Support: support@scichart.com
// Sales:   sales@scichart.com
//
// ColumnDrillDownView.swift is part of the SCICHART® Examples. Permission is hereby granted
// to modify, create derivative works, distribute and publish any part of this source
// code whether for commercial, private or personal use.
//
// The SCICHART® examples are distributed in the hope that they will be useful, but
// without any warranty. It is provided "AS IS" without warranty of any kind, either
// expressed or implied.
//******************************************************************************

// MARK: Custom modifier

class ColumnHitTest: SCIGestureModifier {
    
    var tapedAtIndex: SCIActionBlock_Pint! = nil
    var doubleTaped: SCIActionBlock! = nil
    
    var _location = CGPoint.zero  // position of tap gesture
    var _index : Int32 = 0
    
    override func onTapGesture (_ gesture: UITapGestureRecognizer!, at view: UIView!) -> Bool {
        let location : CGPoint = gesture.location(in: view)
        let rs : SCIRenderSurfaceProtocol = self.parentSurface.renderSurface!
        if (!rs.isPoint(withinBounds: location)) { return false }
        if (gesture.state == .ended) {
            // save location of touch
            _location = location
            let parent: SCIChartSurfaceProtocol = self.parentSurface
            let series = parent.renderableSeries as SCIRenderableSeriesCollection
            let surface = parent.renderSurface!
            let actualLocation = surface.point(inChartFrame: _location)
            
            let count : Int32 = series.count()
            // check every renderable series for hit
            for i in 0..<count {
                let rSeries: SCIRenderableSeriesProtocol = series.item(at:UInt32(i))!
                let data: SCIRenderPassDataProtocol = rSeries.currentRenderPassData
                let hitTest: SCIHitTestProviderProtocol = rSeries.hitTestProvider() // get hit test tools
                
                // hit test verticaly: check if vertical projection through touch location crosses chart
                let hitTestResult : SCIHitTestInfo = hitTest.hitTestVerticalAt(x: Double(actualLocation.x), y: Double(actualLocation.y), radius: 5, onData: data)
                if (hitTestResult.match).boolValue {
                    // if hit is registered on series
                    // get values at closest point to hit test position
                    _index = hitTestResult.index
                    if tapedAtIndex != nil {
                        tapedAtIndex(_index)
                    }
                }
            }
        }
        return true
    }
    
    override func onDoubleTapGesture (_ gesture: UITapGestureRecognizer!, at view: UIView!) -> Bool {
        let location: CGPoint = gesture.location(in: view)
        let rs: SCIRenderSurfaceProtocol = self.parentSurface.renderSurface!
        if (!rs.isPoint(withinBounds: location)) { return false }
        if doubleTaped != nil {
            doubleTaped()
        }
        return true
    }
}

// MARK: Palette provider

class ColumnDrillDownPalette: SCIPaletteProvider {
    var styles = [SCIStyleProtocol?]()
    
    func addStyle(_ style: SCIStyleProtocol) {
        styles.append(style)
    }
    
    override func styleFor(x: Double, y: Double, index: Int32) -> SCIStyleProtocol! {
        if (index > Int32(styles.count)) {
            return nil
        }
        return styles[Int(index)]
    }
}

// MARK: Example view

class ColumnDrillDownView: SingleChartLayout {
    
    var _totalData: SCIXyDataSeries! = nil
    
    var _firstColumn: SCIFastColumnRenderableSeries! = nil
    var _secondColumn: SCIFastColumnRenderableSeries! = nil
    var _thirdColumn: SCIFastColumnRenderableSeries! = nil
    var _totalColumn: SCIFastColumnRenderableSeries! = nil
    
    var _isShowingTotal = false
    
    override func initExample() {
        let xAxis = SCICategoryNumericAxis()
        xAxis.growBy = SCIDoubleRange(min: SCIGeneric(0.1), max: SCIGeneric(0.1))
        
        let yAxis = SCINumericAxis()
        yAxis.growBy = SCIDoubleRange(min: SCIGeneric(0.1), max: SCIGeneric(0.1))
        
        _totalData = SCIXyDataSeries(xType: .int32, yType: .int32)

        _firstColumn = SCIFastColumnRenderableSeries()
        _firstColumn.fillBrushStyle = SCILinearGradientBrushStyle(colorCodeStart: 0xFF00FFFF, finish: 0xA000FFFF, direction: .vertical)
        _firstColumn.strokeStyle = nil
        _firstColumn.dataSeries = getDataSeries()
        
        _secondColumn = SCIFastColumnRenderableSeries()
        _secondColumn.fillBrushStyle = SCILinearGradientBrushStyle(colorCodeStart: 0xFF00FF00, finish: 0xA000FF00, direction: .vertical)
        _secondColumn.strokeStyle = nil
        _secondColumn.dataSeries = getDataSeries()
        
        _thirdColumn = SCIFastColumnRenderableSeries()
        _thirdColumn.fillBrushStyle = SCILinearGradientBrushStyle(colorCodeStart: 0xFFFF0000, finish: 0xA0FF0000, direction: .vertical)
        _thirdColumn.strokeStyle = nil
        _thirdColumn.dataSeries = getDataSeries()

        let palette = ColumnDrillDownPalette()
        palette.addStyle(_firstColumn.style)
        palette.addStyle(_secondColumn.style)
        palette.addStyle(_thirdColumn.style)
        
        _totalColumn = SCIFastColumnRenderableSeries()
        _totalColumn.fillBrushStyle = SCILinearGradientBrushStyle(colorCodeStart: 0xFF505050, finish: 0xA550005, direction: .vertical)
        _totalColumn.strokeStyle = nil
        _totalColumn.dataSeries = _totalData
        _totalColumn.paletteProvider = palette

        let drillDownModifier = ColumnHitTest()
        weak var wSelf = self
        drillDownModifier.doubleTaped = { wSelf?.showSeries((wSelf?._totalColumn)!, isTotal: true) }
        drillDownModifier.tapedAtIndex = { (index ) in wSelf?.showDetailedChart(index) }
        
        SCIUpdateSuspender.usingWithSuspendable(surface) {
            self.surface.xAxes.add(xAxis)
            self.surface.yAxes.add(yAxis)
            self.surface.chartModifiers = SCIChartModifierCollection(childModifiers: [SCIPinchZoomModifier(), drillDownModifier])
            
            let animation = SCIWaveRenderableSeriesAnimation(duration: 3, curveAnimation: .easeOut)
            animation.repeatable = true
            
            self._firstColumn.addAnimation(animation)
            self._secondColumn.addAnimation(animation)
            self._thirdColumn.addAnimation(animation)
            self._totalColumn.addAnimation(animation)
            
            self.showSeries(self._totalColumn, isTotal: true)
        }
    }
    
    fileprivate func getDataSeries() -> SCIXyDataSeries! {
        let dataSeries = SCIXyDataSeries(xType: .int32, yType: .int32)
        
        var total: UInt32 = 0
        let count = arc4random_uniform(5) + 3
        for i in 0..<count {
            let value = arc4random_uniform(5 + i) + 1
            total += value
            dataSeries.appendX(SCIGeneric(i), y: SCIGeneric(value))
        }
        _totalData.appendX(SCIGeneric(3), y: SCIGeneric(total))
        
        return dataSeries
    }
    
    fileprivate func showDetailedChart(_ index: Int32) {
        if (!_isShowingTotal) { return }
        if index == 0 {
            showSeries(_firstColumn, isTotal: false)
        } else if index == 1 {
            showSeries(_secondColumn, isTotal: false)
        } else if index == 2 {
            showSeries(_thirdColumn, isTotal: false)
        }
    }
    
    fileprivate func showSeries(_ series: SCIRenderableSeriesProtocol, isTotal: Bool) {
        _isShowingTotal = isTotal
        surface.renderableSeries.clear()
        surface.renderableSeries.add(series)
        
        surface.viewportManager.zoomExtents()
    }
}
